media-libs/openjpeg: add 2.5.0

laumann opened the pull request:
Also update to EAPI 8, and bump the test data to the latest commit
possible. Drop all security patches from v2.4.0 as they are part of the
upstream release.

Closes: https://bugs.gentoo.org/844064
Bug: https://bugs.gentoo.org/783513
Bug: https://bugs.gentoo.org/484802
Signed-off-by: Thomas Bracht Laumann Jespersen <t@laumann.xyz>

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+
+EAPI=8
+
+CMAKE_ECLASS=cmake
```
This is default w/ EAPI 8+ for cmake-multilib.eclass

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+	sys-libs/zlib:="
+DEPEND="${RDEPEND}"
+BDEPEND="
+	doc? ( app-doc/doxygen )"
```
All on one line for BDEPEND?

laumann replied:
done

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+RESTRICT="!test? ( test )"
+
+RDEPEND="
+	media-libs/lcms:2
```
I don't mind if you prefer blank line but if we're going to do it, should do it at end of RDEPEND string too. I suggest:
RDEPEND="media-libs/lcms:2

laumann replied:
for multi-line dependencies i would maybe tend to prefer blanks in the beginning and end, so that's what I did. I do agree with the consistency in both ends of the string

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+DOCS=( AUTHORS.md CHANGELOG.md NEWS.md README.md THANKS.md )
+
+PATCHES=(
+	"${FILESDIR}/${PN}-2.5.0-gnuinstalldirs.patch" # bug 667150
```
Consistency nit:
```suggestion
	"${FILESDIR}/${PN}-2.5.0-gnuinstalldirs.patch" # bug #25523 667150
```

laumann replied:
is https://bugs.gentoo.org/25523 the bug you meant to link?

thesamesam replied:
sorry, no, I just meant to prefix with #!

laumann replied:
Ah ok :-) done!

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+		-DBUILD_PKGCONFIG_FILES=ON # always build pkgconfig files, bug #539834
+		-DBUILD_TESTING="$(multilib_native_usex test)"
+		-DBUILD_DOC=$(multilib_native_usex doc ON OFF)
+		-DBUILD_CODEC=$(multilib_is_native_abi && echo ON || echo OFF)
```
... wait, can't we just use multilib_native_usex here?

thesamesam replied:
Oops :)

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+
+	# Cheat a little bit and force disabling fixed point magic
+	# The test suite is extremely fragile to small changes
+	# bug 715130, bug 715422
```
Consistency nit:
```suggestion
	# bug #715130, bug #715422
```

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+		myctestargs+=( --extra-verbose --output-on-failure )
+	fi
+
+	echo ctest "${myctestargs[@]}" "$@"
```
edo, maybe

laumann replied:
tried it out, but I'd need to make it `nonfatal` and it prints a scary "Failed to run ctest" message (as an error), but it's only accepted failing tests that actually failed.

laumann replied:
Thinking a little more about this - does it make sense to have `"$@"` here? Is `multilib_src_test()` called with arguments?

laumann replied:
I'll drop it then...

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+
+	echo ""
+
+	awk -F: '{ print $2 }' "${FAILEDTEST_LOG}" > "${FAILURES_LOG}"
```
Missing die.

laumann replied:
Changed it up so the output is piped straight into "while read". There's no intermediate file anymore.

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+
+	echo ""
+	einfo "Note: Upstream is maintaining a list of known test failures."
+	einfo "We will now compare our test results against this list and sort out any known failure."
```
```suggestion
	einfo "We will now compare our test results against this list and remove any known failing tests."
```

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+	fi
+
+	echo ""
+	einfo "Note: Upstream is maintaining a list of known test failures."
```
```suggestion
	einfo "Note: Upstream maintains a list of known test failures."
```

laumann replied:
done

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+}
+
+multilib_src_test() {
+	if ! multilib_is_native_abi ; then
```
I wonder if this is even true. It feels like it.. shouldn't be?

We do it all the time.

thesamesam replied:
I'll merge without this fixed but please look into it for future

laumann replied:
I'm not really sure how to respond here, and I think I don't fully understand what this check is for. I don't mind looking into it before merging though.

thesamesam commented on media-libs/openjpeg/openjpeg-2.5.0.ebuild:
```diff
@@ -0,0 +1,141 @@
+inherit cmake-multilib flag-o-matic
+
+# Make sure that test data are not newer than release;
+# otherwise we will see "Found-But-No-Test" test failures!
```
Optional: maybe add a comment explaining how to do it

laumann replied:
done

laumann commented:
@thesamesam addressed your review comments (and fixed the sourcing of the ebuild). Also changed:

 - use `elog` instead of echo, and remove all but one `echo`
 - rework instances of `cat` - one of them was a cp
 - a few smaller rewrites

laumann commented:
Comparing the image folders after install for 2.4.0-r3 and 2.5.0, i get the following:

```diff
--- image-old	2022-05-16 19:14:03.813839221 +0200
+++ image-new	2022-05-16 19:13:30.033461191 +0200
@@ -2,29 +2,28 @@
 /usr
 /usr/share
 /usr/share/doc
-/usr/share/doc/openjpeg-2.4.0-r3
-/usr/share/doc/openjpeg-2.4.0-r3/THANKS.md.bz2
-/usr/share/doc/openjpeg-2.4.0-r3/NEWS.md.bz2
-/usr/share/doc/openjpeg-2.4.0-r3/README.md.bz2
-/usr/share/doc/openjpeg-2.4.0-r3/AUTHORS.md.bz2
-/usr/share/doc/openjpeg-2.4.0-r3/CHANGELOG.md.bz2
+/usr/share/doc/openjpeg-2.5.0
+/usr/share/doc/openjpeg-2.5.0/THANKS.md.bz2
+/usr/share/doc/openjpeg-2.5.0/AUTHORS.md.bz2
+/usr/share/doc/openjpeg-2.5.0/CHANGELOG.md.bz2
+/usr/share/doc/openjpeg-2.5.0/NEWS.md.bz2
+/usr/share/doc/openjpeg-2.5.0/README.md.bz2
 /usr/bin
 /usr/bin/opj_dump
 /usr/bin/opj_compress
 /usr/bin/opj_decompress
 /usr/include
-/usr/include/openjpeg-2.4
-/usr/include/openjpeg-2.4/opj_stdint.h
-/usr/include/openjpeg-2.4/openjpeg.h
-/usr/include/openjpeg-2.4/opj_config.h
+/usr/include/openjpeg-2.5
+/usr/include/openjpeg-2.5/opj_stdint.h
+/usr/include/openjpeg-2.5/openjpeg.h
+/usr/include/openjpeg-2.5/opj_config.h
 /usr/lib64
 /usr/lib64/libopenjp2.so
 /usr/lib64/libopenjp2.so.7
-/usr/lib64/libopenjp2.so.2.4.0
+/usr/lib64/libopenjp2.so.2.5.0
 /usr/lib64/pkgconfig
 /usr/lib64/pkgconfig/libopenjp2.pc
 /usr/lib64/cmake
-/usr/lib64/cmake/openjpeg-2.4
-/usr/lib64/cmake/openjpeg-2.4/OpenJPEGConfig.cmake
-/usr/lib64/cmake/openjpeg-2.4/OpenJPEGTargets-gentoo.cmake
-/usr/lib64/cmake/openjpeg-2.4/OpenJPEGTargets.cmake
+/usr/lib64/cmake/OpenJPEGConfig.cmake
+/usr/lib64/cmake/OpenJPEGTargets-relwithdebinfo.cmake
+/usr/lib64/cmake/OpenJPEGTargets.cmake
```
which after adjusting for things that just have a new version number looks like it reduces to:

```diff
-/usr/lib64/cmake/openjpeg-2.4
-/usr/lib64/cmake/openjpeg-2.4/OpenJPEGConfig.cmake
-/usr/lib64/cmake/openjpeg-2.4/OpenJPEGTargets-gentoo.cmake
-/usr/lib64/cmake/openjpeg-2.4/OpenJPEGTargets.cmake
+/usr/lib64/cmake/OpenJPEGConfig.cmake
+/usr/lib64/cmake/OpenJPEGTargets-relwithdebinfo.cmake
+/usr/lib64/cmake/OpenJPEGTargets.cmake
```
where things have actually moved and/or gotten new names. The OpenJPEGTargets-gentoo.cmake and OpenJPEGTargets-relwithdebinfo.cmake are the "same" autogenerated files (their contents differ, but the intention appears to be the same). I'll be honest and say I'm not sure if this has any impact (but I suspect that it would) - I'll look into it to see if I overlooked anything.

```diff
# diff -Naur \
#    /var/tmp/portage/media-libs/openjpeg-2.4.0-r3/image/usr/lib64/cmake/openjpeg-2.4/OpenJPEGTargets-gentoo.cmake \
#    /var/tmp/portage/media-libs/openjpeg-2.5.0/image/usr/lib64/cmake/OpenJPEGTargets-relwithdebinfo.cmake
--- /var/tmp/portage/media-libs/openjpeg-2.4.0-r3/image/usr/lib64/cmake/openjpeg-2.4/OpenJPEGTargets-gentoo.cmake	2022-05-16 19:08:30.000000000 +0200
+++ /var/tmp/portage/media-libs/openjpeg-2.5.0/image/usr/lib64/cmake/OpenJPEGTargets-relwithdebinfo.cmake	2022-05-16 19:07:39.000000000 +0200
@@ -1,43 +1,43 @@
 #----------------------------------------------------------------
-# Generated CMake target import file for configuration "Gentoo".
+# Generated CMake target import file for configuration "RelWithDebInfo".
 #----------------------------------------------------------------
 
 # Commands may need to know the format version.
 set(CMAKE_IMPORT_FILE_VERSION 1)
 
-# Import target "openjp2" for configuration "Gentoo"
-set_property(TARGET openjp2 APPEND PROPERTY IMPORTED_CONFIGURATIONS GENTOO)
+# Import target "openjp2" for configuration "RelWithDebInfo"
+set_property(TARGET openjp2 APPEND PROPERTY IMPORTED_CONFIGURATIONS RELWITHDEBINFO)
 set_target_properties(openjp2 PROPERTIES
-  IMPORTED_LINK_INTERFACE_LIBRARIES_GENTOO "m"
-  IMPORTED_LOCATION_GENTOO "${_IMPORT_PREFIX}/lib64/libopenjp2.so.2.4.0"
-  IMPORTED_SONAME_GENTOO "libopenjp2.so.7"
+  IMPORTED_LINK_INTERFACE_LIBRARIES_RELWITHDEBINFO "m"
+  IMPORTED_LOCATION_RELWITHDEBINFO "${_IMPORT_PREFIX}/lib64/libopenjp2.so.2.5.0"
+  IMPORTED_SONAME_RELWITHDEBINFO "libopenjp2.so.7"
   )
 
 list(APPEND _IMPORT_CHECK_TARGETS openjp2 )
-list(APPEND _IMPORT_CHECK_FILES_FOR_openjp2 "${_IMPORT_PREFIX}/lib64/libopenjp2.so.2.4.0" )
+list(APPEND _IMPORT_CHECK_FILES_FOR_openjp2 "${_IMPORT_PREFIX}/lib64/libopenjp2.so.2.5.0" )
 
-# Import target "opj_decompress" for configuration "Gentoo"
-set_property(TARGET opj_decompress APPEND PROPERTY IMPORTED_CONFIGURATIONS GENTOO)
+# Import target "opj_decompress" for configuration "RelWithDebInfo"
+set_property(TARGET opj_decompress APPEND PROPERTY IMPORTED_CONFIGURATIONS RELWITHDEBINFO)
 set_target_properties(opj_decompress PROPERTIES
-  IMPORTED_LOCATION_GENTOO "${_IMPORT_PREFIX}/bin/opj_decompress"
+  IMPORTED_LOCATION_RELWITHDEBINFO "${_IMPORT_PREFIX}/bin/opj_decompress"
   )
 
 list(APPEND _IMPORT_CHECK_TARGETS opj_decompress )
 list(APPEND _IMPORT_CHECK_FILES_FOR_opj_decompress "${_IMPORT_PREFIX}/bin/opj_decompress" )
 
-# Import target "opj_compress" for configuration "Gentoo"
-set_property(TARGET opj_compress APPEND PROPERTY IMPORTED_CONFIGURATIONS GENTOO)
+# Import target "opj_compress" for configuration "RelWithDebInfo"
+set_property(TARGET opj_compress APPEND PROPERTY IMPORTED_CONFIGURATIONS RELWITHDEBINFO)
 set_target_properties(opj_compress PROPERTIES
-  IMPORTED_LOCATION_GENTOO "${_IMPORT_PREFIX}/bin/opj_compress"
+  IMPORTED_LOCATION_RELWITHDEBINFO "${_IMPORT_PREFIX}/bin/opj_compress"
   )
 
 list(APPEND _IMPORT_CHECK_TARGETS opj_compress )
 list(APPEND _IMPORT_CHECK_FILES_FOR_opj_compress "${_IMPORT_PREFIX}/bin/opj_compress" )
 
-# Import target "opj_dump" for configuration "Gentoo"
-set_property(TARGET opj_dump APPEND PROPERTY IMPORTED_CONFIGURATIONS GENTOO)
+# Import target "opj_dump" for configuration "RelWithDebInfo"
+set_property(TARGET opj_dump APPEND PROPERTY IMPORTED_CONFIGURATIONS RELWITHDEBINFO)
 set_target_properties(opj_dump PROPERTIES
-  IMPORTED_LOCATION_GENTOO "${_IMPORT_PREFIX}/bin/opj_dump"
+  IMPORTED_LOCATION_RELWITHDEBINFO "${_IMPORT_PREFIX}/bin/opj_dump"
   )
 
 list(APPEND _IMPORT_CHECK_TARGETS opj_dump )
```
which looks like (1) new version and (2) configuration changed name from "Gentoo" to "RelWithDebInfo".

laumann commented:
Answered my own question re the autogenerated .cmake files - in EAPI 7, the build type is a non-standard "Gentoo" (from cmake.eclass), and in EAPI the default is "RelWithDebInfo".

thesamesam commented:
```
 * Applying openjpeg-2.5.0-gnuinstalldirs.patch ...
/var/tmp/portage/media-libs/openjpeg-2.5.0/temp/environment: line 1144: /var/tmp/portage/media-libs/openjpeg-2.5.0/files/openjpeg-2.5.0-gnuinstalldirs.patch: No such file or directory
/var/tmp/portage/media-libs/openjpeg-2.5.0/temp/environment: line 1147: /var/tmp/portage/media-libs/openjpeg-2.5.0/files/openjpeg-2.5.0-gnuinstalldirs.patch: No such file or directory          [ !! ]
 * ERROR: media-libs/openjpeg-2.5.0::gentoo failed (prepare phase):
 *   patch -p1  failed with /var/tmp/portage/media-libs/openjpeg-2.5.0/files/openjpeg-2.5.0-gnuinstalldirs.patch
 *
```

laumann commented:
> ```
>  * Applying openjpeg-2.5.0-gnuinstalldirs.patch ...
> /var/tmp/portage/media-libs/openjpeg-2.5.0/temp/environment: line 1144: /var/tmp/portage/media-libs/openjpeg-2.5.0/files/openjpeg-2.5.0-gnuinstalldirs.patch: No such file or directory
> /var/tmp/portage/media-libs/openjpeg-2.5.0/temp/environment: line 1147: /var/tmp/portage/media-libs/openjpeg-2.5.0/files/openjpeg-2.5.0-gnuinstalldirs.patch: No such file or directory          [ !! ]
>  * ERROR: media-libs/openjpeg-2.5.0::gentoo failed (prepare phase):
>  *   patch -p1  failed with /var/tmp/portage/media-libs/openjpeg-2.5.0/files/openjpeg-2.5.0-gnuinstalldirs.patch
>  *
> ```

:facepalm:

I'll fix it.
